contracts/0.4.24/lib/StakeLimitUtils.sol

```diff
@@ -0,0 +1,127 @@
+        uint32 growthBlocks = uint32(_slotValue >> MAX_STAKE_LIMIT_GROWTH_BLOCKS_OFFSET);
+        if (growthBlocks > 0) {
+            stakeLimitIncPerBlock = maxStakeLimit / growthBlocks;
+        }
+        prevStakeLimit = uint96(_slotValue >> PREV_STAKE_LIMIT_OFFSET);
+        prevStakeBlockNumber = uint32(_slotValue >> PREV_STAKE_BLOCK_NUMBER_OFFSET);
+    }
+
+    /**
+    * @notice Pack stake limit params and state into a slot.
+    */
+    function encodeStakeLimitSlot(
+        uint256 _maxStakeLimit,
+        uint256 _stakeLimitIncreasePerBlock,
+        uint256 _prevStakeLimit,
+        uint256 _prevStakeBlockNumber
+    ) internal pure returns (uint256 ret) {
+        require(_maxStakeLimit <= uint96(-1), "TOO_LARGE_MAX_STAKE_LIMIT");
+        require(_maxStakeLimit >= _stakeLimitIncreasePerBlock, "TOO_LARGE_LIMIT_INCREASE");
+        require(_prevStakeLimit <= uint96(-1), "TOO_LARGE_PREV_STAKE_LIMIT");
+        require(_prevStakeBlockNumber <= uint32(-1), "TOO_LARGE_BLOCK_NUMBER");
+
+        require(
+            (_stakeLimitIncreasePerBlock == 0)
+            || (_maxStakeLimit / _stakeLimitIncreasePerBlock <= uint32(-1)),
+            "TOO_SMALL_LIMIT_INCREASE"
+        );
+
+        ret = _maxStakeLimit << MAX_STAKE_LIMIT_OFFSET
+            | _prevStakeLimit << PREV_STAKE_LIMIT_OFFSET
+            | _prevStakeBlockNumber << PREV_STAKE_BLOCK_NUMBER_OFFSET;
+
+        if (_stakeLimitIncreasePerBlock > 0) {
+            ret |= (_maxStakeLimit / _stakeLimitIncreasePerBlock) << MAX_STAKE_LIMIT_GROWTH_BLOCKS_OFFSET;
+        }
+    }
+
+    /**
+    * @notice Calculate stake limit for the current block.
+    */
+    function calculateCurrentStakeLimit(uint256 _slotValue) internal view returns(uint256 limit) {
+        (
+            uint256 maxStakeLimit,
+            uint256 stakeLimitIncPerBlock,
+            uint256 prevStakeLimit,
+            uint256 prevStakeBlockNumber
+        ) = decodeStakeLimitSlot(_slotValue);
+
+        limit = prevStakeLimit + ((block.number - prevStakeBlockNumber) * stakeLimitIncPerBlock);
+        if (limit > maxStakeLimit) {
+            limit = maxStakeLimit;
+        }
+    }
+
+    /**
+    * @notice Write new prev stake limit and current block number
+    */
+    function updatePrevStakeLimit(uint256 _slotValue, uint256 _newPrevLimit) internal view returns(uint256) {
+        return (
+            (_slotValue & STAKE_LIMIT_PARAMS_MASK)
+            | _newPrevLimit << PREV_STAKE_LIMIT_OFFSET
+            | block.number << PREV_STAKE_BLOCK_NUMBER_OFFSET
+        );
+    }
+
+    /**
+    * @notice check if staking is on pause (i.e. slot contains zero value)
+    */
+    function isStakingPaused(uint256 _slotValue) internal pure returns(bool) {
+        return (_slotValue == 0);
+    }
+
+    /**
+    * @notice check if rate limit is set (otherwise staking is unlimited)
+    */
+    function isStakingRateLimited(uint256 _slotValue) internal pure returns(bool) {
```

Comment:
Maybe it's better to rename `isStakingLimitApplied` for consistency? Or rename similar variables in Lido.sol to `isStakingRateLimited `. Wdyt?